feat: Middleware capability introduced (#155)
kpn/kstreams: docs/middleware.md

# Middleware

Kstreams allows you to include middlewares for adding behavior to streams. 

> A *middleware* is a `callable` that works with every `ConsumerRecord` (CR) *before* it is processed by a specific `stream`.  Also *after* the `CR` has been handled `Middlewares` also have access to the `stream`, `send` function.

Reviewer: ```suggestion
A *middleware* is a `callable` that works with every `ConsumerRecord` (CR) *before* and *after* is processed by a `stream`. 

`Middlewares` have access to the `stream` itself and the `send` function.
```

- It takes each `CR` that arrives to a topic.
- It can then do something to the `CR` or run any needed code.
- Then it passes the `CR` to be processed by the specific stream.
- Once the `CR` is processed by the stream, the chain is "completed".
- If there is code after the `self.next_call(cr)` then it will be executed.

> 

Reviewer: ```suggestion
Let's say we assign the following Middlewares: `[MiddlewareFoo, MiddlewareBar]`

We can see that `MiddlewareFoo` will be called first, and last:

```mermaid
flowchart LR
    m1i[MiddlewareFoo]
    m2i[MiddlewareBar]
    fn[stream]
    m1o[MiddlewareFoo]
    m2o[MiddlewareBar]
    
    m1i --> m2i
    m2i --> fn
    fn --> m2o
    m2o --> m1o
    ```
```

Reviewer: I would add something like that, maybe better phrased

Author: What you point is the description of `middleware chain`. I think at the beginning of the docs is better to describe the process as simple as possible focus in the ConsumerRecord, `Stream` and  _one_ `middleware`. I will addyour suggestion in the `Middleware chain` section

Author: After discussing we realized that a sequence diagram is a better option
![Screenshot 2024-01-16 at 16 16 51](https://github.com/kpn/kstreams/assets/2828842/c0e7dfe2-34b2-49a1-9d58-5a1233efec54)

        # the next call could be another middleware
        return await self.next_call(cr)
```

Then, we have to include the middleware. It can be per `stream` or per application level (`stream_engine`):

=== "Include middleware in stream"
    ```python
    from kstreams import ConsumerRecord, middleware

    from .engine import stream_engine

    middlewares = [middleware.MiddlewareFactory(ElasticMiddleware)]

    @stream_engine.stream("kstreams-topic", middlewares=middlewares)
        async def processor(cr: ConsumerRecord):
            ...
    ```
=== "Include middleware at application level"
    ```python
    from kstreams import ConsumerRecord, middleware, create

    middlewares = [middleware.MiddlewareFactory(ElasticMiddleware)]
    
    stream_engine = kstreams.create_engine(
        title="my-stream-engine",
        middlewares=middlewares
    )
    ```

!!! note
    The example middlewares make sense to include it at the application level, but other ones like Dead Letter Queue (DLQ) make sense to add them at stream level

!!! note
    The `Middleware` concept also applies for `async generators` (yield from a stream)

## Adding extra configuration to middlewares

> If you want to provide extra configuration to middleware you should override the __init__ method, ensuring that is contains the keywargs `next_call`, `send` and `stream`, then any remaining are optional `keywargs`.

Reviewer: Should we recommend using `super()` instead? Would it work?

```
class DLQMiddleware:
    def __init__(self, topic: str, *args, timeout: int = 10, **kwargs):
        super().__init__(self, *args, **kwargs)
        self.topic = topic
        self.timeout = timeout
```

```

> If you want to provide extra configuration to middleware you should override the __init__ method, ensuring that is contains the keywargs `next_call`, `send` and `stream`, then any remaining are optional `keywargs`.

Reviewer: This line also mentions "override", but the example is implementing the protocol. Should the example inherit from `BaseMiddleware`?

Author: Yes, it is possible and then we can use `super().__init__(next_call=next_call, send=send, stream=stream)`
